src/core/renderer.js

```diff
@@ -144,73 +144,89 @@ class Renderer {
         return 1;
     }
 
-    _computeDrawMetadata(renderLayer) {
-        const tiles = renderLayer.getActiveDataframes();
+    /**
+     * Run aggregation functions over the visible features.
+     */
+    _runViewportAggregations(renderLayer) {
+        const dataframes = renderLayer.getActiveDataframes();
         const viz = renderLayer.viz;
         const aspect = this.getAspect();
-        let drawMetadata = {
-            zoom: 1. / this._zoom,
-            columns: []
-        };
+        const scale = 1. / this._zoom;
 
-        const s = 1. / this._zoom;
-
-        const rootExprs = viz._getRootExpressions();
         // Performance optimization to avoid doing DFS at each feature iteration
-        const viewportExprs = [];
+        const viewportExpressions = this._getViewportExpressions(viz._getRootExpressions());
+
+        if (!viewportExpressions.length) {
+            return;
+        }
+
+        viewportExpressions.forEach(expr => expr._resetViewportAgg());
+
+        // Keep track of features processed in viewPort expressions
+        const processedFeaturesIDs = new Set();
+
+        dataframes.forEach(dataframe => {
+            for (let i = 0; i < dataframe.numFeatures; i++) {
+                // If feature has been acumulated ignore it
+                if (processedFeaturesIDs.has(dataframe.properties.cartodb_id[i])) {
```

Comment:
I think it is better to have a small function self-explanatory than writing a comment, but that is a personal opinion.

Example:

```js
_isFeatureAccumulated(processedFeaturesIDs, dataframe) {
   return processedFeaturesIDs.has(dataframe.properties.cartodb_id[i]);
} 
```

and 

```js
isFeatureInViewport(dataframe, index, scale, aspect) {
  return dataframe.inViewport(i, scale, this._center, aspect):
}
```

Reply:
I disagree for this case, this code is quite performant intensive, I prefer to avoid function calls

Reply:
I don't see a proper way to handle `_isFeatureAccumulated` because the state lives in the scope...

I think the comment is ok to clarify the intention

Reply:
Ok, I will buy you that.